src/main.rs

```diff
@@ -0,0 +1,291 @@
+        self.timer += 1;
+        (self.n, x, self.peak_flag, self.state) // self.peak_flag)
+    }
+    fn update_hr(&mut self, start_n : usize) {
+        // Search for peak in above data
+        if self.above_pts.capacity() > 1 {
+            let mut above_max = 0i32;
+            let mut above_ix = 0usize;
+            for (i, val) in self.above_pts.iter().enumerate() {
+                if above_max < *val {
+                    above_max = *val;
+                    above_ix = i;
+                }
+            }
+            // Given when above_pts started, and above_ix, calc delta to last peak
+            let this_peak_n = start_n + above_ix;
+            let delta_n = this_peak_n - self.last_peak_n;
+            self.last_peak_n = this_peak_n;
+
+            // if delta > 200 && delta < 2000 {
+            self.hr = 60000f64 / delta_n as f64;
+
+            c5412::set_value(self.hr as u32);
+        }
+    }
+    fn hr(&self) -> f64 { self.hr }
+    fn above_below(&self) -> (i32, i32) { // 140us
+        let mut above = 0i32;
+        if self.above_pts.capacity() > 1 {
+            let i = self.above_pts.iter();
+            for ii in i {
+                if above < *ii {
+                    above = *ii;
+                }
+            }
+        }
+        let mut below = 0i32;
+        if self.below_pts.capacity() > 1 {
+            let i = self.below_pts.iter();
+            for ii in i {
+                if below > *ii {
+                    below = *ii;
+                }
+            }
+        }
+        (above, below)
+    }
+    fn new() -> Hr {
+        let yc = 32768i32;
+        Hr {
+            dc_ema: yc as f64,
+            lp_ema: yc as f64,
+            threshold_ema: yc as f64,
+            n : 0usize,
+            state : 0u8,
+            timer : 0usize,
+            peak_flag : 0u8,
+            wild_flag : 0u8,
+            above_pts : ConstGenericRingBuffer::<i32, ABOVE_SIZE>::new(),
+            below_pts : ConstGenericRingBuffer::<i32, BELOW_SIZE>::new(),
+            last_peak_n : 0usize,
+            hr : 0f64,
+        }
+    }
+    fn help(&self) -> (u32, u32) {
+        (self.dc_ema as u32, self.threshold_ema as u32)
+    }
+}
+
+static SAMPLE_SIGNAL: Signal<CriticalSectionRawMutex, u32> = Signal::new();
+
+type UART = embassy_stm32::usart::UartTx<'static, embassy_stm32::peripherals::USART3, embassy_stm32::peripherals::DMA1_CH1>;
+static UART_INST: StaticCell<UART> = StaticCell::new();
+
+type LED1 = embassy_stm32::gpio::Output<'static, embassy_stm32::peripherals::PB0>;
+static LED1_INST: StaticCell<LED1> = StaticCell::new();
+type LED3 = embassy_stm32::gpio::Output<'static, embassy_stm32::peripherals::PB14>;
+static LED3_INST: StaticCell<LED3> = StaticCell::new();
+
+type BUTTON1 = embassy_stm32::gpio::Input<'static, embassy_stm32::peripherals::PC13>;
+static BUTTON1_INST: StaticCell<BUTTON1> = StaticCell::new();
+
+#[embassy_executor::task]
+async fn process(uart_ref: &'static mut UART,
```

Comment:
comment on purpose of LEDs: pulse on led1, and "low pass disabled" on red led3
comment on how uart is outputting raw data because DUMP is on
Rename DUMP to something else?

Reply:
Done